using-python-for-science/intro-to-environments.md

```diff
@@ -0,0 +1,322 @@
+versions on your computer. Probably 95% of installation problems we help people
+with have to do with people not knowing which Python they are using out of the
+multiple versions on their computer.
+
+To this end, here we show you the single most important diagnostic when
+troubleshooting your Python installation problems. In your Python program,
+whether that program is commands entered into the Python prompt, a .py file, or
+a Jupyter notebook, type:
+
+```python
+import sys
+print(sys.executable)
+print(sys.path)
+```
+
+The first print function call prints the location of the *interpreter* running
+your code --- in the email analogy, this is the same as the email program with
+which you are reading or sending email. Most usually, when you install a Python
+package but can't access it, it's because you installed it for a different
+Python interpreter.
+
+Similarly, `sys.path` tells you the folders within your computer where Python
+is looking for programs or libraries to import. If you install a library with
+conda or pip, it *will* end up in one of these folders. If it's not there, or
+it's a different version from the one you expect, it usually means, again, that
+you've installed it into a different environment.
+
+# Managing environments
+
+Hopefully, the above ideas are enough to keep you oriented in the face of
+different versions of Python and different Python environments within your
+machine. When in doubt, don't panic and print `sys.executable`! Now that you
+have a compass, though, let's see how you can create and navigate environments.
+Once you can create, modify, and destroy environments at will, you'll be able
+to marshall the entire Scientific Python ecosystem at will!
+
+In Python, there are two main tools to manage environments: virtualenv, and
+conda. Various other tools, such as pipenv and mamba, give you options to
+manipulate either of those two types of environments. The concepts for all
+these tools are the same, so we will focus on conda and leave you, brave
+reader, to figure out the other tools when necessary, armed with the concepts
+from this chapter. Indeed, a good exercise when you are done with this chapter
+is to uninstall everything, then start from scratch with mamba.
+
+The first thing to do is to get conda. Confusingly, even now we have choices to
+make: we can either install miniconda or Anaconda. Miniconda is *just* conda,
+and what we will install going forward. Anaconda includes miniconda *as well
+as* a whole bunch of Scientific Python packages hand-compiled by the engineers
+at Anaconda, Inc, all neatly bundled into a single, massive environment.
+Although this can be convenient, in our case we want to create, modify, and
+destroy environments at will, so we will skip the One Huge Environment To Rule
+Them All.
+
+As of this writing, you can download miniconda at
+https://anaconda.org/miniconda, but if that changes, simply search the web for
```

Comment:
This URL is actually totally wrong :joy: I'm wondering whether we should point users to miniforge instead? https://github.com/conda-forge/miniforge

Or micromamba: https://mamba.readthedocs.io/en/latest/micromamba.html  <-- but I'm more hesitant about this, much more recent.